src/flash.c

```diff
@@ -5,11 +5,88 @@
  */
 
 #include <flash.h>
+
+#include <log.h>
 #include <target/flash.h>
+#include <private/rom.h>
+
+/* Flash geometry constants */
+#define STUB_FLASH_SECTOR_SIZE          0x1000
+#define STUB_FLASH_BLOCK_SIZE           0x10000
+#define STUB_FLASH_PAGE_SIZE            0x100
+#define STUB_FLASH_STATUS_MASK          0xFFFF
+
+/* ROM */
+
+extern int esp_rom_spiflash_config_param(uint32_t device_id,
+                                         uint32_t chip_size, uint32_t block_size, uint32_t sector_size,
+                                         uint32_t page_size, uint32_t status_mask);
+
+static uint32_t device_id_to_flash_size(uint32_t device_id)
+{
+    STUB_LOG_TRACEF("device_id: 0x%x\n", device_id);
+    uint32_t size = 0;
+    const uint32_t id = device_id & 0xff;
+    switch (id) {
+    case 0x12:
+        size = 256 * 1024;
+        break;
+    case 0x13:
+        size = 512 * 1024;
+        break;
+    case 0x14:
+        size = 1 * 1024 * 1024;
+        break;
+    case 0x15:
+        size = 2 * 1024 * 1024;
+        break;
+    case 0x16:
+        size = 4 * 1024 * 1024;
+        break;
+    case 0x17:
+        size = 8 * 1024 * 1024;
+        break;
+    case 0x18:
+        size = 16 * 1024 * 1024;
+        break;
+    case 0x19:
+        size = 32 * 1024 * 1024;
+        break;
+    case 0x20:
+        size = 64 * 1024 * 1024;
+        break;
+    case 0x21:
+        size = 128 * 1024 * 1024;
+        break;
+    case 0x22:
+        size = 256 * 1024 * 1024;
+        break;
+    case 0x39:
+        size = 32 * 1024 * 1024;
+        break;
+    default:
+        size = 0;
+    }
+    return size;
+}
 
 void stub_lib_flash_init(void **state)
 {
+    STUB_LOG_TRACE();
+    // TODO: check if we don't need the first initialization?
+    // how we can check - only via EXTMEM_ICACHE_ENABLE?
     stub_target_flash_init(state);
+    const uint32_t device_id = stub_target_flash_device_id();
+    const uint32_t flash_size = device_id_to_flash_size(device_id);
```

Comment:
Will we continue even with the unknow flash chip? Better to return error and report it to OpenOCD/esptool

Reply:
will fixed in https://github.com/espressif/esp-stub-lib/pull/16